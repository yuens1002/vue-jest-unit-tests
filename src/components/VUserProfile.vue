<template>
  <div class="user-profile">
    <img class="user-profile__avatar" :src="user.avatar_url" />
    <div class="user-profile__name">
      {{ user.name }}
    </div>
    <div class="user-profile__bio">
      {{ user.bio }}
    </div>
  </div>
</template>

<script>
export default {
  name: 'UserProfile',
  props: {
    user: {
      type: Object,
      required: true,
      default: () => ({})
    },
    aProp: {
      type: Object,
      required: true,
      default: () => ({})
    }
  }
}
</script>

