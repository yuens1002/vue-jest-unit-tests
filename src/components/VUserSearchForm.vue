<template>
  <form @submit.prevent="$emit('submitted', username)">
    <input type="text" v-model="username" />
    <button type="submit">Enviar</button>
  </form>
</template>

<script>
export default {
  name: 'UserSearchForm',
  data () {
    return {
      username: ''
    }
  }
}
</script>
