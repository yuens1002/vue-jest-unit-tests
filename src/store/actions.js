// eslint-disable-next-line eol-last
export default {}
