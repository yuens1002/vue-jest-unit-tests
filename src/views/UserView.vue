<template>
  <div>
    <VUserSearchForm @submitted="searchUser" />
    <VUserProfile :user="user" :aProp="aProp" />
    <p>hello</p>
  </div>
</template>

<script>
import { mapState } from 'vuex'
import VUserProfile from '../components/VUserProfile'
import VUserSearchForm from '../components/VUserSearchForm'
export default {
  name: 'UserView',
  data () {
    return {
      aProp: {
        name: ''
      }
    }
  },
  methods: {
    searchUser (username) {
      this.$store.dispatch('SEARCH_USER', {username})
    }
  },
  computed: {
    ...mapState({
      user: 'user'
    })
  },
  components: {
    VUserProfile,
    VUserSearchForm
  }
}
</script>
