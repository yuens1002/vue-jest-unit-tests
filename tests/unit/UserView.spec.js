/* eslint-disable no-undef */

import { shallowMount, createLocalVue } from '@vue/test-utils'
import Vuex from 'vuex'
import UserView from '@/views/UserView'
import VUserSearchForm from '@/components/VUserSearchForm'
import VUserProfile from '@/components/VUserProfile'
import intialState from '@/store/state'
import userFixture from './fixtures/user'
import actions from '@/store/actions'

//read up on this...
jest.mock('@/store/actions')

const localVue = createLocalVue()
localVue.use(Vuex)

describe('UserView', () => {
  let state
  const build = () => {
    const wrapper = shallowMount(UserView, {
      localVue,
      store: new Vuex.Store({ state, actions })
    })
    return {
      wrapper,
      userSearchForm: () => wrapper.find(VUserSearchForm),
      userProfile: () => wrapper.find(VUserProfile)
    }
  }
  beforeEach(() => {
    jest.resetAllMocks()
    state = { ...intialState }
  })
  it('seraches for a user when received "submitted" action', () => {
    // arrange
    const expectedUser = 'kur'
    const { userSearchForm } = build()

    // act as if userSearchForm emits the "submitted" event
    userSearchForm().vm.$emit('submitted', expectedUser)

    // assert
    expect(actions.SEARCH_USER).toHaveBeenCalled()
    // calls[0] is Vuex object itself, [1] is the payload
    expect(actions.SEARCH_USER.mock.calls[0][1]).toEqual({ username: expectedUser })
  })
  it('renders the component', () => {
    // assert
    expect(build().wrapper.html()).toMatchSnapshot()
  })
  it('renders direct decendents', () => {
    // arrange
    const { userSearchForm, userProfile } = build()

    // assert
    expect(userSearchForm().exists()).toBe(true)
    expect(userProfile().exists()).toBe(true)
  })
  it('passes name prop from UserView to VUserProfile', () => {
    // arrange
    const { userProfile, wrapper } = build()
    // assert
    expect(userProfile().vm.aProp).toBe(wrapper.vm.aProp)
  })
  it('passes user prop from state to VUserProfile', () => {
    state.user = userFixture
    const { userProfile } = build()
    // assert
    expect(userProfile().vm.user).toBe(state.user)
  })
})
