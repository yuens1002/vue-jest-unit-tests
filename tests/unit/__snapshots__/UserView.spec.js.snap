// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserView renders the component 1`] = `
<div>
  <vusersearchform-stub></vusersearchform-stub>
  <vuserprofile-stub user="[object Object]" aprop="[object Object]"></vuserprofile-stub>
  <p>hello</p>
</div>
`;
